docs/util/util-intro.md

```diff
@@ -0,0 +1,107 @@
+[`./spark/src/main/scala/geotrellis/spark/io/file/FileAttributeStore.scala`](https://github.com/geotrellis/geotrellis/blob/master/spark/src/main/scala/geotrellis/spark/io/file/FileAttributeStore.scala)
+for examples of usage.
+
+### MethodExtensions.scala ###
+
+This file contains the `MethodExtensions` trait,
+which is the base trait for all implicit classes containing extension methods.
+
+#### Motivation ####
+
+This mechanism is present to allow a more organized code base,
+it allows methods to be written in an abstract way so that they do not only apply to one particular chain of types,
+it allows functions to be written in an abstract way so that they require extensions rather than types,
+and it provides a canonical extension mechanism for users of the library.
+
+#### Usage ####
+
+Examples of how to use the mechanism provided by the `MethodExtensions` trait
+are pervasive throughout the Geotrellis codebase.
+
+One example is the addition of `RasterRDD` crop methods in
+[`spark/src/main/scala/geotrellis/spark/crop/TileLayerRDDCropMethods.scala`](https://github.com/geotrellis/geotrellis/blob/master/spark/src/main/scala/geotrellis/spark/crop/TileLayerRDDCropMethods.scala),
+made available in [`spark/src/main/scala/geotrellis/spark/crop/Implicits.scala`](https://github.com/geotrellis/geotrellis/blob/master/spark/src/main/scala/geotrellis/spark/crop/Implicits.scala)
+(where the an implicit class with the extension methods is wrapped in a trait),
+and [`spark/src/main/scala/geotrellis/spark/package.scala`](https://github.com/geotrellis/geotrellis/blob/master/spark/src/main/scala/geotrellis/spark/package.scala)
+(where the trait is mixed into the package object via the line `with crop.Implicits`).
+
+Below we present another example, this one using the fictitious `geotrellis.raster.display` namespace,
+to discuss the preferred mechanics for using the `MethodExtensions` functionality.
+First, a `trait` or `abstract class` is created which extends `MethodsExtensions`
+(compare the example below with `TileLayerRDDCropMethods.scala` linked-to above):
+```scala
+package geotrellis.raster.display
+
+import geotrellis.util.MethodExtensions
+
+
+trait DisplayMethods[T <: AnyVal] MethodExtensions[T] {
+  def display(): Unit = println(self)
+}
+```
+
+Next, the implicit class containing the methods is presented as an implicit class within a trait called `Implicit`
+(compare the example below with `Implicits.scala` linked-to above):
+```scala
+package geotrellis.raster.display
+
+
+trait Implicits {
+  implicit class withDisplayMethods[T <: AnyVal](val self: T)
+      extends DisplayMethods[T]
+}
+```
+
+Finally, the `Implicit` trait is mixed into the package object
+(compare the example below with the `package.scala` file linked-to above):
+```scala
+package object raster
+    extends buffer.Implicits
+    with display.Implicits // Add Our Extension Methods
+    with Serializable
+{
+...
+}
+```
```

Comment:
There are couple of caveats that I would want to mention:

- If your method extensions need [ContextBounds](http://docs.scala-lang.org/tutorials/FAQ/context-and-view-bounds.html) or other implicit parameters, you can't define them as a trait. In this case, define them as an `abstract class` and implement the `withFooMethods` in the `Implicits` as you mentioned, extending the abstract class. Example: https://github.com/geotrellis/geotrellis/blob/master/spark/src/main/scala/geotrellis/spark/stitch/StitchRDDMethods.scala#L38

- If you don't require context bounds or other implicit parameters, but you're defining the extension methods on a core type of the library, then look to see if there is already an implicit class that gives method extensions to core types in the package object, and don't define a new implicit class in an Implicits object. Example: https://github.com/geotrellis/geotrellis/blob/master/raster/src/main/scala/geotrellis/raster/package.scala#L70

- If you are going to be working with RDD's, and potentially serializing implicit parameters of the implicit class that will be extending the trait, then you should define an object that contains the functionality, and the methods trait or abstract class should call out to that object. Example: https://github.com/geotrellis/geotrellis/blob/master/spark/src/main/scala/geotrellis/spark/filter/TileLayerRDDFilterMethods.scala#L35. This avoids serialization issues.